lib/release/release.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package release
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/gravitational/roundtrip"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/trace"
+)
+
+type ClientConfig struct {
+	// TLSConfig is the client TLS configuration
+	TLSConfig *tls.Config
+}
+
+func (c *ClientConfig) CheckAndSetDefaults() error {
+	if c.TLSConfig == nil {
+		return trace.BadParameter("missing TLS configuration")
+	}
+
+	return nil
+}
+
+type Client struct {
+	// client is the client used to make calls to the release API
+	client *roundtrip.Client
+}
+
+func NewClient(cfg ClientConfig) (*Client, error) {
+	err := cfg.CheckAndSetDefaults()
+	if err != nil {
+		return &Client{}, trace.Wrap(err)
+	}
+
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: cfg.TLSConfig,
+		},
+	}
+
+	if err != nil {
+		return &Client{}, trace.Wrap(err)
+	}
+
+	client, err := roundtrip.NewClient(fmt.Sprintf("https://%s", cfg.TLSConfig.ServerName), "", roundtrip.HTTPClient(httpClient))
+	if err != nil {
+		return &Client{}, trace.Wrap(err)
+	}
+
+	return &Client{
+		client: client,
+	}, nil
+}
+
+func (c *Client) ListReleases(ctx context.Context) ([]*types.Release, error) {
+	if c.client == nil {
+		return nil, trace.BadParameter("client not initialized")
+	}
+
+	resp, err := c.client.Get(ctx, c.client.Endpoint("teleport-ent"), nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if resp.Code() == http.StatusUnauthorized {
+		return nil, trace.AccessDenied("access denied by the release server")
+	}
+
+	var releases []*types.Release
+	json.Unmarshal(resp.Bytes(), &releases)
+
+	// TODO remove unplublished releases
+
+	// add human-readable display sizes
+	for i := range releases {
+		for j := range releases[i].Assets {
+			releases[i].Assets[j].DisplaySize = byteCount(releases[i].Assets[j].Size_)
```

Comment:
The `Size_` field member is confusing. Could we decouple our API  from the external HTTP third-party API ?

Reply:
I believe `Size_` is generated with the underscore because `Size` might cause a name collision with some internal stuff.

I've added the suggestion in https://github.com/gravitational/teleport/pull/19565/commits/04d4b1f9d1921cfd755a7c82659ccba7545e94ea. In this commit, I've added new types to map to the return of the API, then translate them to the GRPC type. Please let me know what you think